include/swift/SIL/SILConstants.h

```diff
@@ -0,0 +1,246 @@
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This defines an interface to represent SIL level structured constants in an
+// memory efficient way.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_SIL_CONSTANTS_H
+#define SWIFT_SIL_CONSTANTS_H
+
+#include "swift/SIL/SILValue.h"
+
+namespace swift {
+class SingleValueInstruction;
+class SILValue;
+class SILBuilder;
+class SerializedSILLoader;
+
+struct APIntSymbolicValue;
+struct ArraySymbolicValue;
+struct EnumWithPayloadSymbolicValue;
+struct UnknownSymbolicValue;
+
+/// When we fail to constant fold a value, this captures a reason why,
+/// allowing the caller to produce a specific diagnostic.  The "Unknown"
+/// SymbolicValue representation also includes a pointer to the SILNode in
+/// question that was problematic.
+enum class UnknownReason {
+  // TODO: Eliminate the default code, by making classifications for each
+  // failure mode.
+  Default,
+
+  /// The constant expression was too big.  This is reported on a random
+  /// instruction within the constexpr that triggered the issue.
+  TooManyInstructions,
+
+  /// A control flow loop was found.
+  Loop,
+
+  /// Integer overflow detected.
+  Overflow,
+
+  /// Unspecified trap detected.
+  Trap,
+};
+
+/// This is the symbolic value tracked for each SILValue in a scope.  We
+/// support multiple representational forms for the constant node in order to
+/// avoid pointless memory bloat + copying.  This is intended to be a
+/// light-weight POD type we can put in hash tables and pass around by-value.
+///
+/// Internally, this value has multiple ways to represent the same sorts of
+/// symbolic values (e.g. to save memory).  It provides a simpler public
+/// interface though.
+class SymbolicValue {
+private:
+  enum RepresentationKind {
+    /// This symbolic value cannot be determined, carries multiple values
+    /// (i.e., varies dynamically at the top level), or is of some type that
+    /// we cannot analyze and propagate (e.g. NSObject).
+    ///
+    RK_Unknown,
+
+    /// This value is known to be a metatype reference.  The type is stored
+    /// in the "metatype" member.
+    RK_Metatype,
+
+    /// This value is known to be a function reference, e.g. through
+    /// function_ref directly, or a devirtualized method reference.
+    RK_Function,
+
+    /// This value is represented with a bump-pointer allocated APInt.
+    RK_Integer,
+
+    /// This value is represented with an inline integer representation.
+    RK_IntegerInline,
+
+    /// This value is an array, struct, or tuple of constants.  This is
+    /// tracked by the "aggregate" member of the value union.
```

Comment:
@marcrasi  "This value is an array, ..." -> For arrays, we will have a different representation kind, right? Guess array should be removed from this comment.